File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.List;

class DnDCharacter {
        private final int strength;
        private final int dexterity;
        private final int constitution;
        private final int intelligence;
        private final int wisdom;
        private final int charisma;
        private final int hitpoints;
        private final Dice dice;

    DnDCharacter() {
        this.dice = new Dice();
        this.strength = ability();
        this.dexterity = ability();
        this.constitution = ability();
        this.intelligence = ability();
        this.wisdom = ability();
        this.charisma = ability();
        this.hitpoints = modifier(constitution) + 10;
    }

    int ability() {
        List<Integer> rolls = dice.fourRolls();
        int lowestRoll = rolls.get(0);
        for (int roll : rolls) if (roll < lowestRoll) lowestRoll = roll;
        rolls.remove((Integer) lowestRoll);
        return rolls.stream().reduce(0, Integer::sum);
    }

    int modifier(int input) {
        return (int) Math.floor((input - 10.0) / 2.0);
    }

    int getStrength() {
        return strength;
    }

    int getDexterity() {
        return dexterity;
    }

    int getConstitution() {
        return constitution;
    }

    int getIntelligence() {
        return intelligence;
    }

    int getWisdom() {
        return wisdom;
    }

    int getCharisma() {
        return charisma;
    }

    int getHitpoints() {
        return hitpoints;
    }
}

File: go/dnd-character/dnd_character.go
package dndcharacter

import (
	"math"
	"math/rand"
)

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return int(math.Floor(float64((score - 10)) / float64(2)))
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	return roll() + roll() + roll()
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	c := Character{
		Ability(),
		Ability(),
		Ability(),
		Ability(),
		Ability(),
		Ability(),
		10,
	}
	c.Hitpoints += Modifier(c.Constitution)
	return c
}

func roll() int {
	return rand.Intn(6) + 1
}
